### DISCUSSION TITLE
[SOLVED] Why both docker and .deb?

### DISCUSSION THREAD

**Discussion by Stooovie** [elonen/clapshot#45] [Q&A] [answered]
Hi, I cannot get this to run. I'm not a total noob and do run many docker containers as well as Proxmox LXCs but I don't understand how to run Clapshot. Why is there both Docker AND .deb installers? 

The .deb packages have been built without errors but from there the instructions are rather unclear to me. 

1) I installed the server .deb with dpkg. What is the client deb for? Does this not work as a pure web app? 
"Server should be put behind a reverse proxy in production, but can be developed and tested without one." - what does this mean? 

2)  tried running `clapshot-server --url-base=http://192.168.0.152:8080 --data-dir=/videos` as 192.168.0.152 is the server Clapshot runs on. Conecting from another guest fails (192.168.0.152 refused to connect.). SOMETIMES I get the "Server connecting" but on subsequent refresh it's "192.168.0.152 refused to connect." again.

3) I'm able to run the demo when host and guest are the same machine. When adding the URL base as instructed like this:

`docker run --rm -it -p 0.0.0.0:8080:80 -v clapshot-demo:/mnt/clapshot-data/data -e CLAPSHOT_URL_BASE=192.168.0.152:8080/  elonen/clapshot:latest-demo `

I get only "Connecting server" in the browser

Thanks for any tips.

**Comment by elonen** [comment, marked as answer]
(Several different questions here, I'll try to clarify them here a bit more than you probably need for others that may find this post.)

A. Purpose of the packages:

- Server .deb installs a local HTTP/WS server can be configured at `/etc/clapshot-server.conf`. It listens to HTTP and websockets connections, and the config default to 127.0.0.1 only in port 8095. The package also provides `lib/systemd/system/clapshot-server.service` to make it start on boot.
- Client .deb is just a convenience that installs static HTML, JS etc in `/usr/share/clapshot-client/www/` and provides example configs for Nginx. You can check its contents with `dpkg-deb -c clapshot-client_*.deb`.
- Keeping stuff in .deb (both client and server) mainly makes upgrades simpler and cleaner in production. 

B. Docker:
- **Build process** uses Docker to avoid contaminating your local system, as well as standardising the environment. Pretty standard.
- The **demo** Docker image is just that, a demo. Internally, it installs the .debs with some insecure defaults, along with Nginx, ffmpeg and so on, so you can test the system on a laptop without further configurations. I wouldn't recommend using it for hosting in production as it doesn't support HTTP or any kind of advanced authentication, logs stay inside the container etc.

C. Reverse proxy suggestion:
- Since the server .deb only listen on 127.0.0.1 by default, you cannot connect to it from remote servers directly. In production, use Nginx as a reverse proxy to handle authentication and add HTTPS/WSS encryption before it reaches port 8095 plaintext.
- In the demo Docker, notice that`http://127.0.0.1:8080` says 8080, while server's `port = 8095`. This is because the container is running an Nginx on 8080, that relays to the Clapshot server on 8095, while adding toy authentication on top of it.

The reason you are getting connection refused on 2. is likely that you specified url-base as `http://192.168.0.152:8080` instead of :8095. Since you are running the server manually without Nginx in front of it (like the demo Docker does), when you try to connect to :8080, nobody's answering.

When you see "connecting server", the best way to debug your setup is to open javascript console and look where exactly it's trying to connect.

(_related to: #30_)

**Comment by elonen** [comment]
Here's an excerpt from a production `clapshot-server.conf`:

```
[general]
url-base = https://clapshot.example.com
data-dir = /mnt/clapshot-data/data

port = 8095
host = 127.0.0.1
bitrate = 16
```

Notice that url-base doesn't specify a port at all. That's because there's an Nginx relaying both client files and server API calls through standard HTTPS port 443 like so:

```
       location / {
                root /var/www/clapshot-client;
                # ^ this is actually a symlink: /var/www/clapshot-client -> /usr/share/clapshot-client/www
                
                index index.html;
                try_files $uri $uri/ =404;

(...some authn/authz stuff here...)

                location ~* \.(html|json|conf|js|css)$ {
                        expires 4h;
                        add_header Cache-Control "public, no-transform";
                }

                location /videos {
                        alias /mnt/clapshot-data/data/videos;
                }

                location /api {
                        proxy_pass http://127.0.0.1:8095/api;

                        # Also pass along websocket
                        proxy_http_version 1.1;
                        proxy_set_header Upgrade $http_upgrade;
                        proxy_set_header Connection "Upgrade";
                        proxy_set_header Host $host;

                        # Pass authenticated username to backend
                        proxy_set_header X-Remote-User-Id    $http_x_authn_user;
                        proxy_set_header X-Remote-User-Name  $display_name;

                        proxy_set_header X-Real-IP $remote_addr;
                        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

                        # Allow large uploads
                        # (TODO: maybe limit to /api/upload ?)
                        proxy_request_buffering        off;
                        client_body_buffer_size        256K;
                        client_max_body_size           50G;
                }
```

**Comment by Stooovie** [comment]
Perfect. Thank you so much. This gives me enough to soldier on.

**Comment by elonen** [comment]
One more point:

Also check `/etc/clapshot_client.conf`.
Since the client is just HTML + JS served from Nginx directly, it doesn't know where the server is unless you tell it.

Here's an a production conf for client, for completeness:
```
{
    "ws_url": "wss://clapshot.example.com/api/ws",
    "upload_url": "https://clapshot.example.com/api/upload",
    "user_menu_extra_items": [
        { "label": "My Videos", "type": "url", "data": "/" },
        { "label": "Logout", "type": "url", "data": "https://vouch.example.com/logout" }
    ],
  "user_menu_show_basic_auth_logout": false
}
```

Notes:
- This is symlinked for Nginx as: `/var/www/clapshot-client/clapshot_client.conf.json -> /etc/clapshot_client.conf`
- The last two lines in this example are logout menu customizations for this particular deployment that uses Okta for authentication.

**Comment by Stooovie** [comment]
That's too much for me... I tried having both removing 8080 from the base URL in .conf as well as having 8095 as part of the base url, to no avail.

I cannot get to /htadmin at all (that is always refused), even from local IP

I probably won't be able to make this work at this time.

Thanks for your time.

**Reply by elonen** [reply]
Ok. Could you tell a little bit about your prospective use case however?

I just added a note to front page about this not being a drop-in replacement for the various cloud services, but if I get a more clear idea about what people are looking for, maybe I could make some kind of a lighter, immediately useful and less authentication/security oriented Docker image  apart from the demo. 🤔

**Comment by Stooovie** [comment]
It looks good enough for general use with remote reviewing and commenting. We use Trello and manually typed timecodes which sort of works but obviously time-based comments are better. I don't have such output that would warrant monthly subscription of the paid services. Clapshot looks great and I think a lightweight docker compose could do (obviously I am not a developer but I run a lot of self-hosted stuff - Immich seems kinda similar in scope). I don't need a lot of security, simple password auth done with Nginx Proxy Server would be enough.

(to add to previous convo, I get the intricacies of reverse proxies and stuff, but why was I never able to access 192.168.0.XXX/htadmin?)

**Comment by elonen** [comment]
👍 How and where would you ideally have installed and configured it?

More specifically:

1. On some docker-enabled NAS or a Linux server with lots of RAM and a good CPU?
2. For configuring a Docker image, would you prefer some terminal installer that asks for endpoints, bitrates etc, just a combined config file in the local folder exposed to Docker (where videos go), or something else? 
3. Do you need HTTPS? (entailing certificate management.)

About the /admin page, without seeing your exact environment, it’s quite hard for me to debug, but I’ll test  Docker image for any regressions.

Not sure if htadmin would be the best solution for a ready-to-use Docker distribution, though. I literally spent 15 minutes to find something that could work as a demo on how to integrate custom authentication systems.
Do you have any preferences?

**Comment by Stooovie** [comment]
1) I was trying this on my homelab Proxmox LXC (container) with just 2 cores and 2GB RAM and the performance was good enough. I don't really need transcoding, I can render to low-bitrate mp4s for approvals. Trying it on my M1 MBP was obviously much faster but either was fine
2) I think the latter (combined config file) would be enough at launch :) more options can be always added later if needed at all.
3) I don't, no. If I do, Nginx Proxy Manager that I use for all my stuff can take care of that.

**Comment by elonen** [comment]
Just tried the htadmin version (with `docker run --rm -it -p 0.0.0.0:8080:80 -v clapshot-demo:/mnt/clapshot-data/data elonen/clapshot:latest-demo-htadmin`). 

For some reason, when I pasted `http://127.0.0.1:8080/htadmin` to Chrome, it automatically changed it to`http://127.0.0.1/htadmin` which obviously doesn't work. Not sure what's going on there.

Anyway, when I manually edited the address line and added the `:8080` back, htadmin page opened and I could login with `htadmin` / `admin`.

**Reply by Stooovie** [reply]
IIRC I did that too (I know this weird behavior of various browsers) - I have Arc, Safari and Edge open and each one behaves differently :) I'll try Firefox too

EDIT: DISREGARD, this is really the culprit with htadmin. Browsers just deciding to remove the :PORT

**Comment by elonen** [comment]
On my Macbook, Firefox even changed http:// to https:// in addition to the port!
Either something wrong with the demo container's Nginx or htadmin settings, or this is some new browser thing, pushing automatically for more secure connections. 👀

**Reply by Stooovie** [reply]
thats craaazy

rn I just need to expose this docker demo to the internet, but so far no dice. I don't know Nginx proper, I only use Nginx Proxy Manager. 

I'm totally confused where I put the :8095 API port that accessing via my domain requires (console says Firefox can’t establish a connection to the server at ws://192.168.0.152:8080/api/ws)

right now I have this:
- /etc/clapshot-server.conf:
```
[general]
url-base = http://clapshot.MYDOMAIN.cz
data-dir = /mnt/clapshot-data/data

port = 8095
host = 127.0.0.1

debug = false

log = /var/log/clapshot.log
```
run the server with this:
`docker run -e CLAPSHOT_URL_BASE='192.168.0.152:8080/' --rm -it -p 0.0.0.0:8080:80 -v clapshot-demo:/mnt/clapshot-data/data elonen/clapshot:latest-demo-htadmin`

and have this in my NPM config:

![Screenshot 2023-12-13 at 16 26 15](https://github.com/elonen/clapshot/assets/10572373/a3aba501-1609-42a6-a83c-b8dfb2a32c34)

And I see just "Connecting server" in browser

**Reply by elonen** [reply]
If it loads the client page from http://clapshot.MYDOMAIN.cz, the redirect should be correct but client is trying to connect somewhere else. Open the javascript console on your browser and look for `..CONNECTING to WS API: ws://<something>/api/ws`.
If `<something>` is not `clapshot.MYDOMAIN.cz` (I'd guess it's `127.0.0.1:8080`), client config needs to be changed so it knows where to connect (see https://github.com/elonen/clapshot/discussions/45#discussioncomment-7841776).

**Reply by Stooovie** [reply]
yes but changed where exactly? that's where I stumble. In the .conf? Or the docker command? or NPM config (that doesn't work)

**Reply by elonen** [reply]
How did you change the server config? clapshot_client.conf should be right next to it.
The "client config" is really just a json file that you should be able to see at http://clapshot.MYDOMAIN.cz/clapshot_client.conf.json
It's served by Nginx along with the HTML, CSS and JS files.

(Additional note: when you add SSL in NPM config, change ws:// to ws:// in clapshot_client.conf, too.)

**Reply by Stooovie** [reply]
huh, where is it in the linux filesystem? it's not in /etc next to clapshot_server.conf but I AM able to access it via the link... weird!

**Reply by elonen** [reply]
The client .deb installs it as a symlink from /var/www where it's accessible by nginx: 
`/etc/clapshot_client.conf -> /var/www/clapshot-client/clapshot_client.conf.json`

**Reply by Stooovie** [reply]
something's weird. There's no /etc/clapshot_client.conf and /var/www doesn't exist. I only see the file in the docker containers:
```
/var/lib/docker/overlay2/167c0c1e02729c3b5fbbe525dd76dee0e66ee473183ce72861e2603a28afb690/diff/etc/clapshot_client.conf
/var/lib/docker/overlay2/4e43b9f3c022005910676d8dad04e6a2eb3415bcaa45d3afe65e095381178048/diff/etc/clapshot_client.conf
/var/lib/docker/overlay2/41qq6zgxhx9i3kkzi3mjdupjq/diff/etc/clapshot_client.conf
```
I feel like I fucked up somewhere early (sorry for wasting so much of your time)

**Reply by elonen** [reply]
No worries, this gives me good information on how this is used.

I'm also a bit confused about how your setup is done exactly. Can you track down where the index.html comes from exactly?
This is what the client .deb contains:
```
$ dpkg-deb -c clapshot-client_0.5.3_all.deb
drwxr-xr-x root/root         0 2023-02-15 22:33 ./
drwxr-xr-x root/root         0 2023-02-15 22:33 ./etc/
-rw-r--r-- root/root       246 2023-02-15 22:33 ./etc/clapshot_client.conf
drwxr-xr-x root/root         0 2023-02-15 22:33 ./usr/
drwxr-xr-x root/root         0 2023-02-15 22:33 ./usr/share/
drwxr-xr-x root/root         0 2023-02-15 22:33 ./usr/share/clapshot-client/
drwxr-xr-x root/root         0 2023-02-15 22:33 ./usr/share/clapshot-client/www/
-rw-r--r-- root/root     12294 2023-02-15 22:33 ./usr/share/clapshot-client/www/android-chrome-192x192.png
(...)
-rw-r--r-- root/root       433 2023-02-15 22:33 ./usr/share/clapshot-client/www/index.html
-rw-r--r-- root/root       263 2023-02-15 22:33 ./usr/share/clapshot-client/www/site.webmanifest
drwxr-xr-x root/root         0 2023-02-15 22:33 ./usr/share/doc/
drwxr-xr-x root/root         0 2023-02-15 22:33 ./usr/share/doc/clapshot-client/
-rw-r--r-- root/root       707 2023-02-15 22:30 ./usr/share/doc/clapshot-client/changelog.gz
-rw-r--r-- root/root      7725 2022-09-20 21:16 ./usr/share/doc/clapshot-client/copyright
drwxr-xr-x root/root         0 2023-02-15 22:33 ./usr/share/doc/clapshot-client/examples/
-rw-r--r-- root/root      1897 2023-02-15 22:15 ./usr/share/doc/clapshot-client/examples/clapshot+htadmin.nginx.conf
-rw-r--r-- root/root      1799 2023-02-15 22:15 ./usr/share/doc/clapshot-client/examples/clapshot.nginx.conf
drwxr-xr-x root/root         0 2023-02-15 22:33 ./var/
drwxr-xr-x root/root         0 2023-02-15 22:33 ./var/www/
lrwxrwxrwx root/root         0 2023-02-15 22:33 ./usr/share/clapshot-client/www/clapshot_client.conf.json -> /etc/clapshot_client.conf
lrwxrwxrwx root/root         0 2023-02-15 22:33 ./var/www/clapshot-client -> /usr/share/clapshot-client/www
```

**Reply by elonen** [reply]
The last two lines show the double symlinking, ultimately to `/usr/share/clapshot-client/www/clapshot_client.conf.json`.

Since there's Docker in the middle, maybe you are accessing the filesystem in some way that doesn't show the symlinks and you need to edit it directly at `/usr/share/clapshot-client/www/clapshot_client.conf.json` ? 

(EDIT: yeah, this seems pretty likely as you said you don't see `/var/www/clapshot-client` either, and it's also a symlink.)

**Reply by Stooovie** [reply]
As suspected, I'm a complete tool and failed to install the client deb. I only installed the server deb before. Now both are installed and the above folder structure does exist.

BUT I'm still unable to get it to work (still Connecting server). This is now my clapshot_client.conf
{
    "ws_url": "ws://clapshot.MYDOMAIN.cz:8095/api/ws",
    "upload_url": "http://clapshot.MYDOMAIN.cz:8095/api/upload",
    "user_menu_extra_items": [
        { "label": "My Videos", "type": "url", "data": "/"}
    ],
    "user_menu_show_basic_auth_logout": false
}

**Reply by elonen** [reply]
Remove the ports. If I understood correctly, your Nginx proxy is already forwarding internet-facing :80 to <lan-ip>:8080, and Nginx inside the Docker is forwarding it to 127.0.0.1:8095. (inside the docker container).

This means that users' clients should be connecting to `ws://clapshot.MYDOMAIN.cz/api/ws`, as that's what they see over the public web. That's what `clapshot_client.conf` specifies: what address the HTML/Javascript client should be connecting to from end users' browsers.

**Reply by Stooovie** [reply]
I'm at loss. in browser, `http://clapshot.jirifiala.cz/clapshot_client.conf.json` returns

```
{
"ws_url": "ws://192.168.0.152:8095/api/ws",
"upload_url": "http://192.168.0.152:8095/api/upload",
"user_menu_extra_items": [
{ "label": "My Videos", "type": "url", "data": "/"}
],
"user_menu_show_basic_auth_logout": false
}
```
but `nano /usr/share/clapshot-client/www/clapshot_client.conf.json` returns
```
{
    "ws_url": "ws://clapshot.jirifiala.cz/api/ws",
    "upload_url": "http://clapshot.jirifiala.cz/api/upload",
    "user_menu_extra_items": [
        { "label": "My Videos", "type": "url", "data": "/"}
    ],
    "user_menu_show_basic_auth_logout": false
}
```
It's like something is not symlinked but that's over my head.

**Reply by elonen** [reply]
It might be coming from browser cache if you didn't restart the container (or container's Nginx). Try force reload (shift-ctrl-r).

...or you have two containers running, one of which has the edited file and the other the old. And NPM forwarding to the wrong one. :)

**Reply by Stooovie** [reply]
guess I'll probably just start from scratch... it's not in browser cache

**Reply by elonen** [reply]
Are you going to install it on the Proxmox LXC? If so, I could write an example shell script that installs everything on a pristine Bookworm without the Docker in between. That might be useful for other people, too, as documentation.

**Reply by Stooovie** [reply]
I would love that oh so much

**Reply by elonen** [reply]
Ok, I'll drop you a note here when it's done.

**Reply by elonen** [reply]
Here's an installation script:
https://gist.github.com/elonen/80a721f13bb4ec1378765270094ed5d5

If you don't need transcoding (your files are small enough already), also add something like
```
bitrate = 999
```
...to `clapshot-server.conf`.
That means "_don't transcode if bitrate is less than 999 Mbit/s_", which it always is.

**Reply by Stooovie** [reply]
hey! it's working, even from the internet domain! there's one step missing that probably could be done by the script:
```
+ test -e /mnt/clapshot-data
+ echo 'Data directory '\''/mnt/clapshot-data'\'' missing. Please mount/create it and run again.'
Data directory '/mnt/clapshot-data' missing. Please mount/create it and run again.
```
but that's no biggie. I set up NPM to port 80 of my local IP and it worked! In Firefox and Edge only. Safari and Arc for some reason absolutely insist on replacing http with https.

Thank you so much!

**Reply by elonen** [reply]
Glad to hear it! That error message was intentional. If you are running it on PVE LXC, you’ll probably want to add a second mount point at /mnt/clapshot-data/ so you can resize and backup it independently of the root FS.

**Reply by elonen** [reply]
If you’re exposing this to the web, I’d suggest also configuring SSL on NPM (perhaps with Lets Encrypt certificates if NPM supports them). In addition to adding security, it would fix the http->https switching with the browsers.

**Reply by Stooovie** [reply]
I tried to do that but with no success so far (I have it done with multiple other things such as my 3D printer interface so I'm not completely in the dark). NPM does create and use LE certs, yes.

**Reply by Stooovie** [reply]
What's weird is that the Mainsail 3D printer interface for example is served over http but the browser doesn't try to redirect it to https. If I configure CS in the same way, I get this in the console and I see "https" in the URL...

`Mixed Content: The page at '<URL>' was loaded over HTTPS, but requested an insecure resource '<URL>'. This request has been blocked; the content must be served over HTTPS.`

When I try setting it as https in NPM, I only get 502 Bad Gateway.

I do "Empty cache and hard reset" on every attempt. The only way to access CS through the internet is to disable SSL at all (every other thing I have via NPM has a LE cert except this) and set it to http.

**Reply by elonen** [reply]
That probably means it’s trying to connect ws:// instead of wss:// from the client. Edit both the server and client configs to make sure there are https+wss instead of http+ws in the URLs. Otherwise client will think you want it NOT to use SSL when connecting to server.

The reason why you have to set (the same) url-base for the server config, too, is that the server sometimes creates URLs (eg. pointing to the videos) that it send to the client verbatim.

In NPM, it should probably say http, because your proxy is connecting to the Clapshot VM unencrypted, and then encrypting it transparently  for the web.

**Reply by Stooovie** [reply]
...I did as you said (wss instead of ws and https instead of Http in both server and client config) and voila! It finally works as expected. Thank you so much! Do you have a tip jar or something?

**Reply by elonen** [reply]
Great! I don’t, but if you feel you understand how the thing works now - or at least how to configure it - a blog post (or perhaps a howto-video if you’re a tuber) could help those that want a similar setup. I’m a too familiar with the inner workings, so another perspective / way of explaining the process might benefit some people.

**Reply by Stooovie** [reply]
I'm sharing it on Mastodon right now, I don't have a blog or YT per se, but I can write something on a few subreddits for example. It's a good project!

**Comment by azukaar** [comment]
Picking on this comment:

> Since the server .deb only listen on 127.0.0.1 by default, you cannot connect to it from remote servers directly. In production, use Nginx as a reverse proxy to handle authentication and add HTTPS/WSS encryption before it reaches port 8095 plaintext.

The issue is that in Docker, the local connection would not be from 127.0.0.1, because typically to run a proxy you keep the port closed. Instead, it would either be coming to
 - container_name:port (in bridge setup)
 - or docker_ip:port (in host mode setup)

I understand there is a CLAPSHOT_URL_BASE to change that behaviour? But then the issue is that the true URL will be the domain, not the container name/ip the reverse-proxy uses to connect.

Any help to understand how to reach the container from a reverse-proxy container would be appreciated :) 

Also for the part about the Docker not being production ready, I think there's a world where users are running the docker container for ease of management, but still run a full reverse proxy on top of it, to make it production ready, so it's an unusual stance to have to call it "for test only". WHat are your thoughts?

**Reply by elonen** [reply]
@azukaar 
URL base (in both client and server configs) is the address that gets passed to the browser, used for CORS etc, so this should point to your reverse proxy - wherever it's actually running - because that's what you want the Javascript and/or web browser to connect to.

If you need the server to bind to something other than localhost (e.g. a container's non-local IP address), the `--host` command line parameter (or with the .deb, equivalently, the `host` config line in `/etc/clapshot-server.conf`) is supposed to change it.
HOWEVER, I have apparently botched this in the codebase and `--host` option seems to be currently ignored!
This is a bug I'll need to fix (added as issue #55).

In principle though, a multi-container (Docker compose, Kubernetes) production deployment as you described is definitely reasonable. It's just not something I'm doing myself (using LXC instead), so the Docker image config I made is for easy building / testing / demoing.

**Reply by azukaar** [reply]
understood thanks for the detailed response!

**Reply by elonen** [reply]
Issue https://github.com/elonen/clapshot/issues/55 fixed in master, --host will work as intended in the next release (or now if you want to rebuild yourself).

(Release 0.6.0 is almost ready, but I think I'll still add some unit tests for the "basic_folders" Organizer plugin, which is written in Python and more prone to undetected bugs than the Rust-based Server.)
